File: nn.jl
using Random
using LinearAlgebra
using Plots
using JuMP
using Ipopt

# Define the neural network structure
struct NeuralNetwork
    W1::Matrix{Float64}
    b1::Vector{Float64}
    W2::Matrix{Float64}
    b2::Vector{Float64}
end

# Initialize the neural network
function init_network(input_size::Int, hidden_size::Int, output_size::Int)
    W1 = randn(hidden_size, input_size) * sqrt(2 / input_size)
    b1 = zeros(hidden_size)
    W2 = randn(output_size, hidden_size) * sqrt(2 / hidden_size)
    b2 = zeros(output_size)
    return NeuralNetwork(W1, b1, W2, b2)
end

# Activation function (ReLU)
relu(x) = max.(0, x)

# Forward pass
function forward(nn::NeuralNetwork, X::Matrix{Float64})
    h = relu.(nn.W1 * X .+ nn.b1)
    y = nn.W2 * h .+ nn.b2
    return y
end

# Loss function (Mean Squared Error)
mse_loss(y_pred::Matrix{Float64}, y_true::Matrix{Float64}) = sum((y_pred .- y_true).^2) / size(y_true, 2)

# Generate synthetic data
function generate_data(n_samples::Int)
    X = randn(2, n_samples)
    y = sin.(X[1, :]) .* cos.(X[2, :])
    return X, reshape(y, 1, :)  # Ensure y is a 1xn_samples matrix
end

# Train using SGD
function train_sgd!(nn::NeuralNetwork, X::Matrix{Float64}, y::Matrix{Float64}, lr::Float64, epochs::Int)
    n_samples = size(X, 2)
    losses = Float64[]
    
    for epoch in 1:epochs
        # Forward pass
        h = relu.(nn.W1 * X .+ nn.b1)
        y_pred = nn.W2 * h .+ nn.b2
        
        # Compute loss
        loss = mse_loss(y_pred, y)
        push!(losses, loss)
        
        # Backward pass
        dy = 2 .* (y_pred .- y) ./ n_samples
        dW2 = dy * h'
        db2 = sum(dy, dims=2)
        dh = nn.W2' * dy
        dh[h .<= 0] .= 0
        dW1 = dh * X'
        db1 = sum(dh, dims=2)
        
        # Update parameters
        nn.W1 .-= lr .* dW1
        nn.b1 .-= lr .* vec(db1)
        nn.W2 .-= lr .* dW2
        nn.b2 .-= lr .* vec(db2)
    end
    
    return losses
end

# Optimize using Ipopt
function optimize_ipopt(X::Matrix{Float64}, y::Matrix{Float64}, hidden_size::Int)
    input_size, n_samples = size(X)
    output_size = size(y, 1)
    
    model = Model(Ipopt.Optimizer)
    
    @variable(model, W1[1:hidden_size, 1:input_size])
    @variable(model, b1[1:hidden_size])
    @variable(model, W2[1:output_size, 1:hidden_size])
    @variable(model, b2[1:output_size])
    
    @variable(model, h[1:hidden_size, 1:n_samples])
    @variable(model, y_pred[1:output_size, 1:n_samples])
    
    @constraint(model, [i=1:hidden_size, j=1:n_samples], h[i,j] >= W1[i,:]' * X[:,j] + b1[i])
    @constraint(model, [i=1:hidden_size, j=1:n_samples], h[i,j] >= 0)
    @constraint(model, [i=1:hidden_size, j=1:n_samples], h[i,j] * (h[i,j] - (W1[i,:]' * X[:,j] + b1[i])) <= 0)
    
    @constraint(model, [i=1:output_size, j=1:n_samples], y_pred[i,j] == W2[i,:]' * h[:,j] + b2[i])
    
    @objective(model, Min, sum((y_pred[i,j] - y[i,j])^2 for i in 1:output_size, j in 1:n_samples) / n_samples)
    
    optimize!(model)
    
    return value.(W1), value.(b1), value.(W2), value.(b2)
end

# Main
function main()
    # Generate data
    X, y = generate_data(1000)
    train_size = 800
    X_train, y_train = X[:, 1:train_size], y[:, 1:train_size]
    X_test, y_test = X[:, (train_size+1):end], y[:, (train_size+1):end]
    
    # Initialize network
    input_size, hidden_size, output_size = 2, 10, 1
    nn_sgd = init_network(input_size, hidden_size, output_size)
    
    # Train using SGD
    sgd_losses = train_sgd!(nn_sgd, X_train, y_train, 0.01, 1000)
    
    # Optimize using Ipopt
    W1, b1, W2, b2 = optimize_ipopt(X_train, y_train, hidden_size)
    nn_ipopt = NeuralNetwork(W1, vec(b1), W2, vec(b2))
    
    # Evaluate both models
    y_pred_sgd = forward(nn_sgd, X_test)
    y_pred_ipopt = forward(nn_ipopt, X_test)
    
    loss_sgd = mse_loss(y_pred_sgd, y_test)
    loss_ipopt = mse_loss(y_pred_ipopt, y_test)
    
    println("SGD Loss: ", loss_sgd)
    println("Ipopt Loss: ", loss_ipopt)
    
    # Plot loss surface
    p = plot(1:length(sgd_losses), sgd_losses, label="SGD", xlabel="Epochs", ylabel="Loss", title="Loss Surface")
    hline!([loss_ipopt], label="Ipopt", linestyle=:dash)
    display(p)
    savefig(p, "loss_surface.png")
end

main()

File: nn.py
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as datasets
import matplotlib.pyplot as plt
import numpy as np
from scipy.optimize import minimize
import multiprocessing as mp
from tqdm import tqdm  # Import tqdm for progress bars

# Set random seed for reproducibility
torch.manual_seed(42)

# Check for GPU
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

# Hyperparameters
input_size = 28 * 28  # 784
hidden_size = 128
num_classes = 10
num_epochs = 5
batch_size = 32
learning_rate = 0.001

# Load the MNIST dataset
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5,), (0.5,))
])

train_dataset = datasets.MNIST(root='./data', train=True, transform=transform, download=True)
test_dataset = datasets.MNIST(root='./data', train=False, transform=transform)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, batch_size=batch_size, shuffle=True)
test_loader = torch.utils.data.DataLoader(dataset=test_dataset, batch_size=batch_size, shuffle=False)

# Define the neural network model
class NeuralNet(nn.Module):
    def __init__(self):
        super(NeuralNet, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size, num_classes)

    def forward(self, x):
        x = x.view(-1, input_size)  # Flatten the input
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        return x

# Function to initialize weights
def initialize_weights(model, zero_weights=False):
    if zero_weights:
        for param in model.parameters():
            nn.init.constant_(param, 0.0)
    else:
        for param in model.parameters():
            nn.init.normal_(param, mean=0.0, std=0.01)

# Evaluating the model
def evaluate_model(model, data_loader):
    model.eval()
    correct = 0
    total = 0

    with torch.no_grad():
        for images, labels in tqdm(data_loader, desc="Evaluating Model", leave=False):
            images = images.to(device)
            labels = labels.to(device)
            outputs = model(images)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    test_accuracy = correct / total
    return test_accuracy

# Objective function for optimization
def objective(params, model_class):
    model = model_class().to(device)  # Move model to GPU
    idx = 0
    for param in model.parameters():
        numel = param.numel()
        param.data = torch.tensor(params[idx:idx + numel], dtype=torch.float32, device=device).view(param.size())
        idx += numel

    # Calculate loss on a small subset for optimization
    loss = 0.0
    for images, labels in train_loader:
        images = images.to(device)
        labels = labels.to(device)
        outputs = model(images)
        loss += nn.CrossEntropyLoss()(outputs, labels).item()
    return loss / len(train_loader)

# Worker function for parallel optimization
def worker(initial_params):
    model_class = NeuralNet
    result = minimize(lambda p: objective(p, model_class), initial_params, method='trust-constr')
    return result.fun, result.x

# Random Search with IPOPT optimization in parallel
def random_search_with_ipopt(num_iterations=10):
    best_loss = float('inf')
    best_model = None

    for _ in tqdm(range(num_iterations), desc="Optimizing with IPOPT"):
        model = NeuralNet().to(device)
        # Random initialization
        initial_params = torch.cat([param.data.view(-1) for param in model.parameters()]).cpu().numpy()

        # Use multiprocessing to optimize in parallel
        with mp.Pool(processes=mp.cpu_count()) as pool:
            results = pool.map(worker, [initial_params] * mp.cpu_count())

        # Evaluate results
        for loss, params in results:
            if loss < best_loss:
                best_loss = loss
                best_model = NeuralNet().to(device)
                idx = 0
                for param in best_model.parameters():
                    numel = param.numel()
                    param.data = torch.tensor(params[idx:idx + numel], dtype=torch.float32, device=device).view(param.size())
                    idx += numel

    return best_model

# Plot training history with progress bar
def plot_training_history(losses, accuracies):
    plt.figure(figsize=(12, 4))

    # Plot loss
    plt.subplot(1, 2, 1)
    plt.plot(losses, label='Train Loss')
    plt.title('Training Loss')
    plt.xlabel('Epochs')
    plt.ylabel('Loss')
    plt.legend()

    # Plot accuracy
    plt.subplot(1, 2, 2)
    plt.plot(accuracies, label='Train Accuracy')
    plt.title('Training Accuracy')
    plt.xlabel('Epochs')
    plt.ylabel('Accuracy')
    plt.legend()

    plt.show()

# Training function with progress bar
def train_model(zero_weights=False):
    model = NeuralNet().to(device)
    initialize_weights(model, zero_weights)

    criterion = nn.CrossEntropyLoss()
    optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

    train_losses = []
    train_accuracies = []

    for epoch in tqdm(range(num_epochs), desc="Training Model"):
        model.train()
        total_loss = 0
        correct = 0
        total = 0

        for images, labels in tqdm(train_loader, leave=False):
            optimizer.zero_grad()  # Zero the gradients
            images = images.to(device)
            labels = labels.to(device)
            outputs = model(images)  # Forward pass
            loss = criterion(outputs, labels)  # Calculate loss
            loss.backward()  # Backward pass
            optimizer.step()  # Update weights
            
            total_loss += loss.item()
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

        avg_loss = total_loss / len(train_loader)
        accuracy = correct / total
        train_losses.append(avg_loss)
        train_accuracies.append(accuracy)

        print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {avg_loss:.4f}, Accuracy: {accuracy:.4f}')

    return train_losses, train_accuracies, model

# Benchmarking all three methods
if __name__ == "__main__":
    print("Training with Zero Weights:")
    losses_zero, accuracies_zero, model_zero = train_model(zero_weights=True)
    evaluate_model(model_zero, test_loader)
    plot_training_history(losses_zero, accuracies_zero)

    print("Training with Random Weights:")
    losses_random, accuracies_random, model_random = train_model(zero_weights=False)
    evaluate_model(model_random, test_loader)
    plot_training_history(losses_random, accuracies_random)

    print("Training with Random Search and IPOPT Optimization:")
    model_ipopt = random_search_with_ipopt()
    test_accuracy = evaluate_model(model_ipopt, test_loader)
    print(f'Test Accuracy (Random Search + IPOPT): {test_accuracy * 100:.2f}%')
